### PR TITLE
add debian directory

### PR DESCRIPTION
**Pull request by PengTian0**
Add debian directory to create rackhd.deb

### CODE REVIEW

**Comment by rackhdmirror** [comment]
@RackHD/corecommitters please review.

**Comment by PengTian0** [comment]
@panpan0000 @changev

**Comment by leoyjchang** [comment]
@PengTian0 I have a general question whether this change could support Ubuntu 16.04 as well? I see some place in script specified for 14.01 (ex: eth1). If not, this should not block this PR but at least we are aware of the situation and can support it later.

**Comment by PengTian0** [comment]
@leoyjchang I suppose there will be a new branch for RackHD and on-xxx to support Ubuntu 16.04.
On this branch, RackHD will change scripts to support Ubuntu 16.04.
Or , I can try to add a config file and debian/scripts to load parameters from the config file~

**Comment by PengTian0** [comment]
@RackHD/corecommitters

File: debian/compat [hunk 1]
```diff
@@ -0,0 +1 @@
+9
```
Comment (panpan0000) [inline debian/compat]: is this file required in source control ?

File: debian/config [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+# Source debconf library.
+. /usr/share/debconf/confmodule
+
+# Please specific the name of control NIC port of RackHD(example : eth1)  > 
+db_input high RackHD/NIC_NAME || true
+db_input high RackHD/WILLING_TO_FORCE_IP || true
+db_go || true
+db_stop
+exit 0
```
Comment (panpan0000) [inline debian/config]: why exit 0 at the end ?

File: debian/postinst [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+check_NIC $RET
+
+update_dhcp_config
+copy_RackHD_static_bins
+enable_services
+create_RackHD_config
+reload_services
+print_usage
+
+db_stop
```
Comment (panpan0000) [inline debian/postinst]: db_stop as soon as user-interaction and variable using completed ?

File: debian/compat (hunk 1 above)
Comment (PengTian0) [inline debian/compat]: It is not required.
But we will get warning without it:

_No compatibility level specified in debian/compat
This package will soon FTBFS; time to fix it!_

File: debian/config (hunk 2 above)
Comment (PengTian0) [inline debian/config]: Thanks for your catch!
I will remove it~

File: debian/postinst (hunk 3 above)
Comment (PengTian0) [inline debian/postinst]: Yes, I will move it after check_NIC

**Comment by PengTian0** [comment]
@RackHD/corecommitters

**Comment by PengTian0** [comment]
@anhou @yyscamper

**Comment by anhou** [comment]
:+1:

File: debian/preinst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+check_nodejs(){
+    installed=0;
+    if [ "`type -p nodejs`" != "" ]; then
+        # Node JS being installed
+        NODEJS_VER=$(nodejs --version | sed s/v//)
+        if [ "$NODEJS_VER" \< "4.0" ] ; then
+            # Remove Old NodeJS
+            echo "[INFO] removing old version [$NODEJS_VER] of NodeJS..."
+            sudo apt-get -y remove nodeijs nodejs-legacy
```
Comment (yyscamper) [inline debian/preinst]: typo: `nodeijs` --> `nodejs`

Comment (PengTian0) [inline debian/preinst]: Thank you!
I will submit a commit to fix it!

File: debian/postinst [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    for i in ${SERVICES[@]}; do
+        echo "start service ${i}"
+        service ${i} stop
+        service ${i} start 
+    done
+}
+
+
+print_usage(){
+    echo "curl http://localhost:8080/api/common/nodes"
```
Comment (PengTian0) [inline debian/postinst]: I will remove usage.

File: debian/changelog [hunk 6]
```diff
@@ -0,0 +1,5 @@
+rackhd (1.1-1) unstable; urgency=low
```
Comment (changev) [inline debian/changelog]: when unreleased, I recommend this to be
````
rackhd (1.1.1) UNRELEASED; urgency=low

Comment (PengTian0) [inline debian/changelog]: We are going to release 2.x and 1.1-1 should be released(unstable).

**Comment by PengTian0** [comment]
@anhou @yyscamper

File: debian/preinst [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            DISTRO="$(lsb_release -s -c)"
+            echo "deb https://deb.nodesource.com/$VERSION $DISTRO main" | sudo tee /etc/apt/sources.list.d/nodesource.list
+            echo "deb-src https://deb.nodesource.com/$VERSION $DISTRO main" | sudo tee -a /etc/apt/sources.list.d/nodesource.list
+
+            sudo apt-get update
+            sudo apt-get -y install nodejs
+            # Install NodeJS 4.x =========== Ends ============
+    fi
+
+    # Double Check NodeJs 2.x installed
```
Comment (anhou) [inline debian/preinst]: NodeJS 2.x ?

Comment (PengTian0) [inline debian/preinst]: sorry, I will fix that!

**Comment by PengTian0** [comment]
@anhou @yyscamper

File: debian/preinst [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+#############################################
+#  Check NodeJS is greater than 4.x
+#############################################
+check_nodejs(){
+    installed=0;
+    if [ "`type -p nodejs`" != "" ]; then
+        # Node JS being installed
+        NODEJS_VER=$(nodejs --version | sed s/v//)
+        if [ "$NODEJS_VER" \< "4.0" ] ; then
```
Comment (yyscamper) [inline debian/preinst]: Currently, we only support nodejs 4.x, we don't promise RackHD works for the version 5.x, 6.x, ...

File: debian/postinst [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+SERVICES=("on-http" "on-taskgraph" "on-dhcp-proxy" "on-tftp" "on-syslog", "rackhd")
+
+#############################################
+# Check if you have 2 NIC as below:
+#  eth0 for the public network - providing access to RackHD APIs, and providing routed (layer3) access to out of band network for machines under management
+#  eth1 for dhcp/pxe to boot/configure the machines. and IP configurated as static ( 172.31.128.0/22 )
+# Parameter #1 ($1) is the NIC name of the control port for RacKHD
+#############################################
+check_NIC(){
```
Comment (yyscamper) [inline debian/postinst]: To be strictly, the network config, isc-dhcp-server config (event http/tftp static files downloading) is not the responsibility for RackHD debian package. It belongs to the deployment guide. Not every customer use isc-dhcp-server, not every customer prefers such kind of network config.

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
Jenkins: test this please
